p2p/keepalive/protocol.go

```diff
@@ -0,0 +1,159 @@
+package keepalive
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"github.com/spacemeshos/go-spacemesh/p2p/p2pcrypto"
+	"github.com/spacemeshos/go-spacemesh/p2p/server"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+)
+
+//Name is the name of the protocol
+const Name = "tcp/v1/keepalive"
+
+//MessageBufSize is the buffer size for the messages channel
+const MessageBufSize = 1000
+
+// KeepAlivePing is the ping protocol ID
+const KeepAlivePing = 0
+
+// DurationBetweenPings determines how long we need to wait between ping messages
+const DurationBetweenPings = time.Second * 10
+
+type baseNetwork interface {
+	RegisterDirectProtocolWithChannel(protocol string, ingressChannel chan service.DirectMessage) chan service.DirectMessage
+	SendWrappedMessage(nodeID p2pcrypto.PublicKey, protocol string, payload *service.DataMsgWrapper) error
+	SubscribePeerEvents() (conn, disc chan p2pcrypto.PublicKey)
+	Disconnect(peer p2pcrypto.PublicKey)
+}
+
+type protocol struct {
+	local                 *node.Info
+	connectPeerChannel    chan p2pcrypto.PublicKey
+	disconnectPeerChannel chan p2pcrypto.PublicKey
+	msgServer             *server.MessageServer
+	logger                log.Log
+	peersLock             sync.RWMutex
+	peers                 map[p2pcrypto.PublicKey]struct{}
+	net                   baseNetwork
+}
+
+func newProtocol(peerChan chan p2pcrypto.PublicKey, svc baseNetwork, log log.Log,
+	messageTimeout time.Duration) *protocol {
+	s := server.NewMsgServer(svc, Name, messageTimeout, make(chan service.DirectMessage, MessageBufSize), log)
+	conn, disc := svc.SubscribePeerEvents()
+	p := &protocol{
+		connectPeerChannel:    conn,
+		disconnectPeerChannel: disc,
+		msgServer:             s,
+	}
+	p.msgServer.RegisterMsgHandler(KeepAlivePing, p.handleKeepAlivePingRequest())
+
+	//spin up a thread that loops on the input from the two channels
+	go p.listenForPeers()
```

Comment:
We try to separate the starting of maintenance routines to different methods this way it's easier to control the order of events. it'll probably be the best to refactor these out to a `Start` method.